# Preprocessing

In [1]:
import pandas as pd
import numpy as np

In [2]:
from functions import *

### Import the datasets

In [3]:
# Open the raw datasets
# Heavy Metal
hm_df = pd.read_excel('data/AllHMFiles.xlsx', engine='openpyxl').fillna(0.0)

# Pesticides
pesticides_df = pd.read_excel('data/AllPesticidesFiles.xlsx', engine='openpyxl').fillna(0.0)
classe_pesticides_df = pd.read_excel('data/pesticides.xlsx', engine='openpyxl')

# Surfaces
surfaces_df = pd.read_excel('data/absSurfs.xlsx', engine='openpyxl')

# Distances
distances_df = pd.read_excel('data/distsOneSheet.xlsx', engine='openpyxl')

In [4]:
hm_df.head(5)

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [5]:
pesticides_df.head(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Nicosulfuron  \
0           0.0           0.0          0.0        0.0  ...           0.0   
1           0.0           0.0          0.0        0.0  ...           0.0   
2           0.0           0.0          0.0        0.0  ...           0.0   
3           0.0           0.0          0.0        0.0  ...           0.0   
4           0.0           0.0          0.0        0.0  ...           0.0   

   Piperonylbutoxide  Prosulfuron  Simazine  Terbuthylazine  Dieldrin  \
0                0.0          0.0       0.0             0.0       0.0   
1                0.0          0.0       0.0             0.0       0.0   
2                0.0          0.0       0.0             0.0       0.0   
3                0.0          0.0       0.0             0.0       0.0   
4                0.0          0.0       0.0             0.0       0.0   

   Fenamidone  Trifluralin  Cycloate  Pentachloroanisol  
0         0.0          0.0       0.0                0.0  
1         0.0          0.0       0.0                0.0  
2         0.0          0.0       0.0                0.0  
3         0.0          0.0       0.0                0.0  
4         0.0          0.0       0.0                0.0  

[5 rows x 117 columns]

In [6]:
classe_pesticides_df.head(5)

importName           name    statusEN   LMR  \
0      Aclonifen      aclonifen  authorized  0.05   
1         Aldrin         aldrin  prohibited  0.01   
2   Ametoctradin   ametoctradin  authorized  0.05   
3  Anthraquinone  anthraquinone  prohibited  0.02   
4   Azoxystrobin   azoxystrobin  authorized  0.05   

                          familyEN          typeEN  
0                   diphenyl ether       herbicide  
1                   organochlorine     insecticide  
2               triazolopyrimidine       fungicide  
3  polycyclic aromatic hydrocarbon  bird repellant  
4                      strobilurin       fungicide

In [7]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [8]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

### Heavy Metal

In [9]:
hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  
0     0.032    40     0.0  
1     0.029    33     0.0  
2     0.038    37     0.0  
3     0.031    37     0.0  
4     0.052    35     0.0

In [10]:
MRL_HM_dict = {
    'Lead' : [1.5],
    'Cadmium' : [1],
    'Arsenic' : [0.3],
    'Mercury' : [1],
    'Copper' : [30]
}

MRL_HM = pd.DataFrame.from_dict(MRL_HM_dict)
MRL_HM

Lead  Cadmium  Arsenic  Mercury  Copper
0   1.5        1      0.3        1      30

In [11]:
hm_df['mrl_lead'] = np.where(hm_df['Lead'] > 1.5, 1, 0)
hm_df['mrl_cadmium'] = np.where(hm_df['Cadmium'] > 1, 1, 0)
hm_df['mrl_arsenic'] = np.where(hm_df['Arsenic'] > 0.3, 1, 0)
hm_df['mrl_mercury'] = np.where(hm_df['Mercury'] > 1, 1, 0)
hm_df['mrl_copper'] = np.where(hm_df['Copper'] > 30, 1, 0)

hm_df.head()

REF....SUBSTANCE   PERIOD   Lead  Cadmium  Arsenic  Mercury  Copper  \
0            B0017  2017/P1  0.081     0.81      0.0      0.0     5.5   
1            B0019  2017/P1  0.062     0.36      0.0      0.0     8.5   
2            B0021  2017/P1  0.150     0.61      0.0      0.0     7.0   
3            B0023  2017/P1  0.067     0.35      0.0      0.0     6.9   
4            B0037  2017/P1  0.120     0.07      0.0      0.0     5.6   

   Chromium  Zinc  Chrome  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
0     0.032    40     0.0         0            0            0            0   
1     0.029    33     0.0         0            0            0            0   
2     0.038    37     0.0         0            0            0            0   
3     0.031    37     0.0         0            0            0            0   
4     0.052    35     0.0         0            0            0            0   

   mrl_copper  
0           0  
1           0  
2           0  
3           0  
4           0

In [12]:
# Dropping columns of metals without specific threshhold level defined by laws
hm_df = hm_df.drop(['Chromium'], axis=1)
hm_df = hm_df.drop(['Chrome'], axis=1)
hm_df = hm_df.drop(['Zinc'], axis=1)

# Dropping other columns not needed anymore
hm_df = hm_df.drop(['Lead'], axis=1)
hm_df = hm_df.drop(['Cadmium'], axis=1)
hm_df = hm_df.drop(['Arsenic'], axis=1)
hm_df = hm_df.drop(['Mercury'], axis=1)
hm_df = hm_df.drop(['Copper'], axis=1)

hm_df.head()

REF....SUBSTANCE   PERIOD  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  \
0            B0017  2017/P1         0            0            0            0   
1            B0019  2017/P1         0            0            0            0   
2            B0021  2017/P1         0            0            0            0   
3            B0023  2017/P1         0            0            0            0   
4            B0037  2017/P1         0            0            0            0   

   mrl_copper  
0           0  
1           0  
2           0  
3           0  
4           0

##### Aleternate dataset (grooup by) for HM

In [13]:
group_hm_df = hm_df.groupby('REF....SUBSTANCE').agg(max).drop(columns=['PERIOD'])
group_hm_df.head()

mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  mrl_copper
REF....SUBSTANCE                                                             
B0012                    0            0            0            0           0
B0017                    0            0            0            0           0
B0019                    0            0            0            0           0
B0021                    0            1            0            0           0
B0023                    0            1            0            0           0

##### Alternate df, with categorical

In [14]:
cat_hm_df = group_hm_df.copy(deep=True)

In [15]:
cat_hm_df['cat'] = cat_hm_df.apply(lambda x: (str(x['mrl_lead']) + 
                                              str(x['mrl_cadmium']) + 
                                              str(x['mrl_arsenic']) +
                                              str(x['mrl_mercury']) + 
                                              str(x['mrl_copper']))  , axis=1)

In [16]:
cat_hm_df['cat'] = cat_hm_df['cat'].astype('category')
cat_hm_df['cat_label'] = cat_hm_df['cat'].cat.codes

labels = cat_hm_df.set_index('cat_label')['cat'].to_dict()
labels

{0: '00000', 3: '01000', 4: '10000', 2: '00100', 1: '00001'}

In [17]:
cat_hm_df = cat_hm_df.drop(columns=['mrl_lead', 
                                    'mrl_cadmium', 
                                    'mrl_arsenic', 
                                    'mrl_mercury', 
                                    'mrl_copper',
                                    'cat'])

In [18]:
cat_hm_df.sample(5)

cat_label
REF....SUBSTANCE           
B0045                     0
F0019                     0
B0039                     0
F0061                     0
FE0001                    0

### Pesticides

Fungicide :

In [19]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].head(5)

importName                       name    statusEN   LMR  \
2         Ametoctradin               ametoctradin  authorized  0.05   
4         Azoxystrobin               azoxystrobin  authorized  0.05   
5            Benalaxyl                  benalaxyl  prohibited  0.05   
7  Benthiavalicarb.iso  benthiavalicarb-isopropyl  authorized  0.05   
8     Benzovindiflupyr           benzovindiflupyr  authorized  0.05   

             familyEN     typeEN  
2  triazolopyrimidine  fungicide  
4         strobilurin  fungicide  
5        phenylamides  fungicide  
7           carbamate  fungicide  
8               amide  fungicide

In [20]:
fungicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'fungicide'].importName.to_list()
fungicides.append('Diphenylamine')
fungicides.append('Tebuconazole')
fungicides

['Ametoctradin',
 'Azoxystrobin',
 'Benalaxyl',
 'Benthiavalicarb.iso',
 'Benzovindiflupyr',
 'Boscalid',
 'Bupirimate',
 'Captan',
 'Carbendazim',
 'Chlorothalonil',
 'Cyazofamid',
 'Cyflufenamid',
 'Cymoxanil',
 'Cyprodinil',
 'Difenoconazole',
 'Dimethomorph',
 'Dimoxystrobin',
 'Dodine',
 'Epoxiconazole',
 'Ethirimol',
 'Famoxadone',
 'Fenamidone',
 'Fenbuconazole',
 'Fenhexamid',
 'Fenpropidin',
 'Fenpropimorph',
 'Fluazinam',
 'Fludioxonil',
 'Fluopicolide',
 'Fluopyram',
 'Flutolanil',
 'Fluxapyroxad',
 'Hexaconazole',
 'Iprovalicarb',
 'Isofetamid',
 'Mandipropamid',
 'Metalaxyl',
 'Metconazole',
 'Metrafenone',
 'Penconazole',
 'Pencycuron',
 'Pentachloroanisol',
 'Prochloraz',
 'Propamocarb',
 'Propiconazole',
 'Prothioconazole_d',
 'Pyraclostrobin',
 'Pyrimethanil',
 'Quinoxyfen',
 'Spiroxamine',
 'Tetraconazole',
 'Thiabendazole',
 'Thiophanate.methyl',
 'Trifloxystrobin',
 'X2.phenylphenol',
 'Zoxamide',
 'Diphenylamine',
 'Tebuconazole']

Insecticide :

In [21]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].head(5)

importName                 name          statusEN   LMR  \
1                Aldrin               aldrin        prohibited  0.01   
13  Chlorantraniliprole  chlorantraniliprole        authorized  0.05   
17  Chlorpyrifos.methyl  chlorpyrifos-methyl        prohibited  0.05   
19            Coumaphos            coumaphos        prohibited  0.10   
20         Cyanofenphos         cyanofenphos  not yet assessed   NaN   

               familyEN       typeEN  
1        organochlorine  insecticide  
13  anthranilic diamide  insecticide  
17      organophosphate  insecticide  
19        thiophosphate  insecticide  
20      organophosphate  insecticide

In [22]:
insecticides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'insecticide'].importName.to_list()
insecticides.append('Diphenylamine')
insecticides.append('Methiocarb_total')
insecticides

['Aldrin',
 'Chlorantraniliprole',
 'Chlorpyrifos.methyl',
 'Coumaphos',
 'Cyanofenphos',
 'Cyflumetofen',
 'Cypermethrin',
 'Deet',
 'Dieldrin',
 'Etofenprox',
 'Fenoxycarb',
 'Fenpropathrin',
 'Flonicamid',
 'Flucythrinate',
 'Flupyradifurone',
 'Imidacloprid',
 'Indoxacarb',
 'Methoxyfenozide',
 'Phosmet',
 'Piperonylbutoxide',
 'Pirimicarb',
 'Spinosad',
 'Spirotetramat',
 'Tau.fluvalinate',
 'Tebufenozide',
 'Tetradifon',
 'Thiacloprid',
 'Diphenylamine',
 'Methiocarb_total']

Herbicide : 

In [23]:
classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].head(5)

importName           name    statusEN   LMR          familyEN  \
0       Aclonifen      aclonifen  authorized  0.05    diphenyl ether   
6     Benfluralin    benfluralin  authorized  0.05    dinitroaniline   
15  Chlorotoluron  chlorotoluron  authorized  0.05        phenylurea   
16   Chlorpropham   chlorpropham  prohibited  0.05         carbamate   
18      Clethodim      clethodim  authorized  0.05  cyclohexanedione   

       typeEN  
0   herbicide  
6   herbicide  
15  herbicide  
16  herbicide  
18  herbicide

In [24]:
herbicides = classe_pesticides_df[classe_pesticides_df['typeEN'] == 'herbicide'].importName.to_list()
herbicides

['Aclonifen',
 'Benfluralin',
 'Chlorotoluron',
 'Chlorpropham',
 'Clethodim',
 'Cycloate',
 'Diflufenican',
 'Ethofumesate',
 'Fluazifop',
 'Flufenacet',
 'Metamitron',
 'Methabenzthiazuron',
 'Metobromuron',
 'Metolachlor',
 'Nicosulfuron',
 'Pendimethalin',
 'Phenmedipham',
 'Propanil',
 'Propyzamide',
 'Prosulfocarb',
 'Prosulfuron',
 'Simazine',
 'Terbuthylazine',
 'Tri.allate',
 'Trifluralin']

Others :

In [25]:
classe_pesticides_df[(classe_pesticides_df['typeEN'] != 'fungicide') & 
                     (classe_pesticides_df['typeEN'] != 'insecticide') & 
                     (classe_pesticides_df['typeEN'] != 'herbicide')]

importName           name    statusEN   LMR  \
3           Anthraquinone  anthraquinone  prohibited  0.02   
34          Diphenylamine  diphenylamine  prohibited  0.05   
69       Methiocarb_total     methiocarb  prohibited  0.05   
86             Propargite     propargite  prohibited  0.05   
100          Tebuconazole   tebuconazole  authorized  0.05   
112  X6.benzylaminopurine  benzyladenine  authorized   NaN   
114                   NaN            NaN         NaN   NaN   

                            familyEN  \
3    polycyclic aromatic hydrocarbon   
34                             amine   
69                         carbamate   
86                               NaN   
100                         triazole   
112                       cytokinins   
114                              NaN   

                                           typeEN  
3                                  bird repellant  
34    growth regulator, fungicide and insecticide  
69   insecticide, molluscicide and bird repellent  
86                                      acaricide  
100                fungicide and growth regulator  
112                              growth regulator  
114                                           NaN

##### Choice to treat the others

Adding to the corresponding categories, stay in "other" ? What about the growth regulator ?

#### Add some columns

In [26]:
pesticides_df['qtt_fungicides'] = pesticides_df.apply(qtt_pesticides, args = fungicides, axis = 1)
pesticides_df['qtt_insecticides'] = pesticides_df.apply(qtt_pesticides, args = insecticides, axis = 1)
pesticides_df['qtt_herbicides'] = pesticides_df.apply(qtt_pesticides, args = herbicides, axis = 1)

In [27]:
pesticides_df['nb_fungicides'] = pesticides_df.apply(nb_pesticides, args = fungicides, axis = 1)
pesticides_df['nb_insecticides'] = pesticides_df.apply(nb_pesticides, args = insecticides, axis = 1)
pesticides_df['nb_herbicides'] = pesticides_df.apply(nb_pesticides, args = herbicides, axis = 1)

In [28]:
pesticides_df['mrl_fungicides'] = pesticides_df.apply(lmr_pesticides, args = fungicides, axis = 1)
pesticides_df['mrl_insecticides'] = pesticides_df.apply(lmr_pesticides, args = insecticides, axis = 1)
pesticides_df['mrl_herbicides'] = pesticides_df.apply(lmr_pesticides, args = herbicides, axis = 1)

In [29]:
pesticides_df.head(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
0            B0017  2017/P1     0.0     0.0000          0.00           0.0   
1            B0019  2017/P1     0.0     0.0188          0.00           0.0   
2            B0021  2017/P1     0.0     0.0711          0.00           0.0   
3            B0023  2017/P1     0.0     0.0294          0.00           0.0   
4            B0037  2017/P1     0.0     0.0000          0.01           0.0   
5            B0039  2017/P1     0.0     0.0000          0.00           0.0   
6            B0041  2017/P1     0.0     0.0000          0.00           0.0   
7            B0043  2017/P1     0.0     0.0000          0.00           0.0   
8            B0045  2017/P1     0.0     0.0000          0.00           0.0   
9            B0059  2017/P1     0.0     0.0000          0.00           0.0   

   Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  Pentachloroanisol  \
0          0.00           0.0         0.00        0.0  ...                0.0   
1          0.00           0.0         0.00        0.0  ...                0.0   
2          0.00           0.0         0.00        0.0  ...                0.0   
3          0.00           0.0         0.00        0.0  ...                0.0   
4          0.00           0.0         0.00        0.0  ...                0.0   
5          0.01           0.0         0.00        0.0  ...                0.0   
6          0.00           0.0         0.00        0.0  ...                0.0   
7          0.00           0.0         0.01        0.0  ...                0.0   
8          0.00           0.0         0.00        0.0  ...                0.0   
9          0.00           0.0         0.00        0.0  ...                0.0   

   qtt_fungicides  qtt_insecticides  qtt_herbicides  nb_fungicides  \
0          0.0000            0.0000            0.00              0   
1          0.0188            0.0000            0.00              1   
2          0.0711            0.0000            0.00              1   
3          0.0294            0.0000            0.00              1   
4          0.0000            0.0000            0.01              0   
5          0.0000            0.0208            0.00              0   
6          0.0000            0.0000            0.00              0   
7          0.0100            0.0000            0.00              1   
8          0.0000            0.0000            0.00              0   
9          0.0145            0.0000            0.00              1   

   nb_insecticides  nb_herbicides  mrl_fungicides  mrl_insecticides  \
0                0              0               0                 0   
1                0              0               0                 0   
2                0              0               1                 0   
3                0              0               0                 0   
4                0              1               0                 0   
5                2              0               0                 0   
6                0              0               0                 0   
7                0              0               0                 0   
8                0              0               0                 0   
9                0              0               0                 0   

   mrl_herbicides  
0               0  
1               0  
2               0  
3               0  
4               0  
5               0  
6               0  
7               0  
8               0  
9               0  

[10 rows x 126 columns]

In [30]:
pesticides_df['year'] = pesticides_df['PERIOD'].apply(lambda x: x[:4])
pesticides_df['trimester'] = pesticides_df['PERIOD'].apply(lambda x: x[-1:])
pesticides_df.sample(10)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
229            F0033  2019/P1     0.0     0.0000           0.0          0.00   
277            B0017  2019/P3     0.0     0.0000           0.0          0.00   
506            F0019  2020/P4     0.0     0.0000           0.0          0.00   
168            F0015  2018/P3     0.0     0.0285           0.0          0.00   
123            B0105  2018/P2     0.0     0.0000           0.0          0.00   
105            F0035  2018/P1     0.0     0.0000           0.0          0.00   
17             F0007  2017/P1     0.0     0.0100           0.0          0.01   
378            F0057  2020/P1     0.0     0.0000           0.0          0.00   
438            B0021  2020/P3     0.0     0.0000           0.0          0.00   
287            B0059  2019/P3     0.0     0.0000           0.0          0.00   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
229           0.0           0.0          0.0        0.0  ...   
277           0.0           0.0          0.0        0.0  ...   
506           0.0           0.0          0.0        0.0  ...   
168           0.0           0.0          0.0        0.0  ...   
123           0.0           0.0          0.0        0.0  ...   
105           0.0           0.0          0.0        0.0  ...   
17            0.0           0.0          0.0        0.0  ...   
378           0.0           0.0          0.0        0.0  ...   
438           0.0           0.0          0.0        0.0  ...   
287           0.0           0.0          0.0        0.0  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
229             0.000             0.0              0                0   
277             0.000             0.0              0                0   
506             0.000             0.0              0                0   
168             0.000             0.0              1                0   
123             0.000             0.0              0                0   
105             0.000             0.0              0                0   
17              0.010             0.0              1                1   
378             0.025             0.0              3                1   
438             0.000             0.0              1                0   
287             0.000             0.0              0                0   

     nb_herbicides  mrl_fungicides  mrl_insecticides  mrl_herbicides  year  \
229              0               0                 0               0  2019   
277              0               0                 0               0  2019   
506              0               0                 0               0  2020   
168              0               0                 0               0  2018   
123              0               0                 0               0  2018   
105              0               0                 0               0  2018   
17               0               0                 0               0  2017   
378              0               1                 0               0  2020   
438              0               0                 0               0  2020   
287              0               0                 0               0  2019   

     trimester  
229          1  
277          3  
506          4  
168          3  
123          2  
105          1  
17           1  
378          1  
438          3  
287          3  

[10 rows x 128 columns]

##### Alternate (smaller) dataset for pesticides

In [31]:
pesti_names = pesticides_df.columns.to_list()[1:-11]

In [32]:
small_pesticides_df = pesticides_df.drop(pesti_names, axis=1)
small_pesticides_df.head(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
0            B0017          0.0000               0.0            0.00   
1            B0019          0.0188               0.0            0.00   
2            B0021          0.0711               0.0            0.00   
3            B0023          0.0294               0.0            0.00   
4            B0037          0.0000               0.0            0.01   

   nb_fungicides  nb_insecticides  nb_herbicides  mrl_fungicides  \
0              0                0              0               0   
1              1                0              0               0   
2              1                0              0               1   
3              1                0              0               0   
4              0                0              1               0   

   mrl_insecticides  mrl_herbicides  year trimester  
0                 0               0  2017         1  
1                 0               0  2017         1  
2                 0               0  2017         1  
3                 0               0  2017         1  
4                 0               0  2017         1

##### Alternate (group by) dataset for pesticides

In [33]:
group_pesti_df = pesticides_df.groupby('REF....SUBSTANCE').agg(max)

In [34]:
group_pesti_df = group_pesti_df.loc[:,'mrl_fungicides' : 'mrl_herbicides']
group_pesti_df.sample(10)

mrl_fungicides  mrl_insecticides  mrl_herbicides
REF....SUBSTANCE                                                  
F0055                          1                 0               0
F0023                          0                 0               1
F0065                          0                 0               0
B0116                          0                 0               0
B0105                          1                 1               1
B0039                          1                 0               0
B0043                          1                 0               1
F0020                          1                 0               0
F0017                          0                 0               0
F0021                          1                 0               0

##### Alternate df, with categorical

In [35]:
cat_pesti_df = group_pesti_df.copy(deep=True)

In [36]:
cat_pesti_df['cat'] = cat_pesti_df.apply(lambda x: (str(x['mrl_fungicides']) + 
                                                    str(x['mrl_insecticides']) + 
                                                    str(x['mrl_herbicides']))  , axis=1)

In [37]:
cat_pesti_df.sample(10)

mrl_fungicides  mrl_insecticides  mrl_herbicides  cat
REF....SUBSTANCE                                                       
F0039                          1                 0               0  100
B0045                          1                 0               0  100
B0059                          1                 0               0  100
B0041                          1                 0               0  100
F0020                          1                 0               0  100
B0163                          1                 0               0  100
B0077                          1                 1               0  110
B0032                          1                 0               0  100
B0043                          1                 0               1  101
B0023                          0                 0               0  000

In [38]:
cat_pesti_df['cat'] = cat_pesti_df['cat'].astype('category')
cat_pesti_df['cat_label'] = cat_pesti_df['cat'].cat.codes

In [39]:
labels = cat_pesti_df.set_index('cat_label')['cat'].to_dict()
labels

{3: '100', 0: '000', 5: '110', 4: '101', 6: '111', 2: '010', 1: '001'}

In [40]:
cat_pesti_df = cat_pesti_df.drop(columns=['mrl_fungicides', 
                                          'mrl_insecticides', 
                                          'mrl_herbicides', 
                                          'cat'])

In [41]:
cat_pesti_df.sample(10)

cat_label
REF....SUBSTANCE           
F0061                     3
B0067                     3
B0023                     0
B0019                     0
F0003                     3
F0071                     3
F0063                     3
F0015                     3
B0039                     3
F0019                     3

### Sites

##### Getting the area for each zone

In [42]:
distances_df.head(5)

Site     polyID  classCLC         dist
0  F0069   FR-88090       231  1183.022362
1  F0069  FR-131773       242   289.926324
2  F0069   FR-46167       211   221.344153
3  F0069   FR-11047       112  1038.383069
4  F0069  FR-195343       311   724.045683

In [43]:
surfaces_df.head(5)

Site  111           112           121            122  123  124  131  132  \
0  F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1  F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2  F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3  B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4  F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   

   133  ...  411  412  421  422  423  511            512  521  522  523  
0  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4  0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  

[5 rows x 45 columns]

In [44]:
# surfaces_df['country'] = surfaces_df['Site'].apply(lambda x: 'FR' if x.startswith('F') else 'BE')
# surfaces_df['country'].value_counts()

In [45]:
sites_df = distances_df.copy(deep=True)

In [46]:
sites_df['surface'] = sites_df.apply(get_surface, axis = 1)

In [47]:
sites_df.sample(10)

Site     polyID  classCLC         dist       surface
126  F0019   FR-30725       122   788.697525  2.638114e+05
339  F0029   FR-24954       112   832.256855  2.943795e+05
530  F0013   FR-81317       231   662.045074  2.151188e+06
114  B0012    BE-3165       112   466.705006  3.756405e+06
535  F0013  FR-158678       243   420.039493  2.507248e+05
91   B0142    BE-4483       122   141.969198  6.365980e+05
97   F0037  FR-265900       331  1286.781067  1.740115e+05
130  F0019  FR-205991       311  1438.808331  7.538446e+05
266  B0059    BE-4271       121   694.110957  3.705107e+05
340  F0029   FR-32866       131   392.579026  2.951948e+05

In [48]:
sites_df['country'] = sites_df['polyID'].apply(lambda x: 'FR' if x.startswith('FR') else 'BE')
sites_df.sample(10)

Site     polyID  classCLC         dist       surface country
236  B0043    BE-8297       211     0.000000  2.969686e+06      BE
361  F0033  FR-224562       312  1433.047518  1.818608e+06      FR
346  F0029  FR-147657       242  1391.076933  7.740918e+05      FR
152  F0075   FR-59263       211     0.000000  4.935220e+06      FR
423  B0019   BE-24778       412   462.433522  3.786370e+05      BE
115  B0012    BE-4014       121     0.000000  2.068585e+06      BE
498  F0061   FR-18801       112    14.450700  2.735732e+06      FR
316  F0017   FR-48732       211  1265.434621  9.580098e+05      FR
97   F0037  FR-265900       331  1286.781067  1.740115e+05      FR
407  B0017   BE-23792       313  1125.759078  1.871540e+06      BE

In [49]:
sites_df['country'].value_counts()

FR    335
BE    297
Name: country, dtype: int64

In [50]:
sites_df = sites_df.drop(['polyID'], axis=1)
sites_df.sample(5)

Site  classCLC         dist       surface country
102  F0037       112   707.582072  1.111996e+06      FR
353  F0033       112  1276.557328  2.810975e+04      FR
579  F0011       242     0.000000  1.525860e+06      FR
74   B0061       311   472.979441  8.684424e+05      BE
204  B0045       331   825.676022  6.126407e+05      BE

##### Get one dataset with : beehives and area/distance for each zone

In [51]:
surfaces_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [52]:
classCLC = surfaces_df.columns.to_list()[1:]

In [53]:
by_sites_df = surfaces_df.copy(deep=True)
by_sites_df

Site  111           112           121            122  123  124  131  132  \
0   F0069  0.0  1.064122e+06  3.587087e+05       0.000000  0.0  0.0  0.0    0   
1   F0025  0.0  0.000000e+00  0.000000e+00  412752.692383  0.0  0.0  0.0    0   
2   F0027  0.0  2.741746e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
3   B0128  0.0  2.470089e+06  4.053755e+06  497496.251770  0.0  0.0  0.0    0   
4   F0023  0.0  0.000000e+00  2.559254e+05       0.000000  0.0  0.0  0.0    0   
..    ...  ...           ...           ...            ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00       0.000000  0.0  0.0  0.0    0   
59  B0069  0.0  8.704243e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
60  B0152  0.0  4.575279e+06  5.726391e+05  201434.000214  0.0  0.0  0.0    0   
61  F0003  0.0  3.425116e+05  0.000000e+00       0.000000  0.0  0.0  0.0    0   
62  B0158  0.0  1.467398e+05  0.000000e+00  318178.187439  0.0  0.0  0.0    0   

    133  ...  411  412  421  422  423  511            512  521  522  523  
0   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
1   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
2   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
3   0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
4   0.0  ...    0  0.0    0    0  0.0  0.0  284728.392578    0    0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...            ...  ...  ...  ...  
58  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
59  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
60  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
61  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  
62  0.0  ...    0  0.0    0    0  0.0  0.0       0.000000    0    0  0.0  

[63 rows x 45 columns]

In [54]:
for CLC in classCLC:
    by_sites_df[CLC] = by_sites_df.apply(distance_area, args = (CLC,), axis = 1)

In [55]:
by_sites_df

Site  111           112           121         122  123  124  131  132  \
0   F0069  0.0  1.442780e+03  3.587087e+05    0.000000  0.0  0.0  0.0  0.0   
1   F0025  0.0  0.000000e+00  0.000000e+00  700.801763  0.0  0.0  0.0  0.0   
2   F0027  0.0  2.604161e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
3   B0128  0.0  2.412962e+03  4.053755e+06  523.436355  0.0  0.0  0.0  0.0   
4   F0023  0.0  0.000000e+00  7.907522e+02    0.000000  0.0  0.0  0.0  0.0   
..    ...  ...           ...           ...         ...  ...  ...  ...  ...   
58  F0007  0.0  0.000000e+00  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
59  B0069  0.0  7.328884e+03  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
60  B0152  0.0  4.575279e+06  5.227880e+02  421.402885  0.0  0.0  0.0  0.0   
61  F0003  0.0  3.859675e+02  0.000000e+00    0.000000  0.0  0.0  0.0  0.0   
62  B0158  0.0  1.574508e+02  0.000000e+00  392.706036  0.0  0.0  0.0  0.0   

    133  ...  411  412  421  422  423  511         512  521  522  523  
0   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
1   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
2   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
3   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
4   0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  463.344615  0.0  0.0  0.0  
..  ...  ...  ...  ...  ...  ...  ...  ...         ...  ...  ...  ...  
58  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
59  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
60  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
61  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  
62  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  

[63 rows x 45 columns]

check if correct

In [56]:
distance = distances_df[(distances_df['Site'] == 'F0023') & 
                        (distances_df['classCLC'] == 512)]['dist'].values
distance

array([614.50674791])

In [57]:
area = surfaces_df[surfaces_df['Site'] == 'F0023']['512'].values[0]
area

284728.392578125

In [58]:
area/distance[0]

463.3446150839585

##### Drop some colomns that don't contain data

In [59]:
by_sites_df = by_sites_df.drop(['132','212','213','222', '223', '241', '244',
              '321', '323', '332', '333', '334', '335', '411',
              '421', '422', '521', '522', '523'], axis=1)

### Merging the datasets into two

3 datasets to 2 : sites + (heavy metal || pesticides)

##### Heavy Metal

In [60]:
hm_df.sample(5)

REF....SUBSTANCE   PERIOD  mrl_lead  mrl_cadmium  mrl_arsenic  \
414            B0067  2020/P4         0            0            0   
344            B0152  2020/P2         0            0            0   
110            F0005  2018/P2         0            0            0   
267            B0067  2019/P4         0            1            0   
289            B0045  2020/P1         0            0            0   

     mrl_mercury  mrl_copper  
414            0           0  
344            0           0  
110            0           0  
267            0           0  
289            0           0

In [61]:
group_hm_df.sample(5)

mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  mrl_copper
REF....SUBSTANCE                                                             
F0037                    1            0            0            0           0
B0039                    0            0            0            0           0
F0073                    0            0            0            0           0
F0029                    0            0            0            0           0
B0021                    0            1            0            0           0

In [62]:
cat_hm_df.sample(5)

cat_label
REF....SUBSTANCE           
B0067                     3
F0065                     1
B0021                     3
B0017                     0
F0023                     0

##### Pesticides

full

In [63]:
pesticides_df.sample(5)

REF....SUBSTANCE   PERIOD  Aldrin  Benalaxyl  Chlorpropham  Cyanofenphos  \
422            F0031  2020/P2     0.0        0.0           0.0           0.0   
465            F0017  2020/P3     0.0        0.0           0.0           0.0   
169            F0017  2018/P3     0.0        0.0           0.0           0.0   
419            F0021  2020/P2     0.0        0.0           0.0           0.0   
358            B0158  2020/P1     0.0        0.0           0.0           0.0   

     Cypermethrin  Ethofumesate  Fludioxonil  Fluopyram  ...  \
422           0.0           0.0          0.0     0.0000  ...   
465           0.0           0.0          0.0     0.0000  ...   
169           0.0           0.0          0.0     0.0000  ...   
419           0.0           0.0          0.0     0.0000  ...   
358           0.0           0.0          0.0     0.0174  ...   

     qtt_insecticides  qtt_herbicides  nb_fungicides  nb_insecticides  \
422            0.0930          0.0000              1                1   
465            0.0000          0.0000              0                0   
169            0.0000          0.0000              0                0   
419            0.0000          0.0000              2                0   
358            0.0069          0.0169              3                1   

     nb_herbicides  mrl_fungicides  mrl_insecticides  mrl_herbicides  year  \
422              0               1                 1               0  2020   
465              0               0                 0               0  2020   
169              0               0                 0               0  2018   
419              0               0                 0               0  2020   
358              2               0                 0               0  2020   

     trimester  
422          2  
465          3  
169          3  
419          2  
358          1  

[5 rows x 128 columns]

In [64]:
small_pesticides_df.sample(5)

REF....SUBSTANCE  qtt_fungicides  qtt_insecticides  qtt_herbicides  \
51             B0039         13.8057            0.0145           0.000   
229            F0033          0.0000            0.0000           0.000   
41             F0009          0.0000            0.0000           0.016   
151            B0043          0.0528            0.0000           0.000   
262            F0013          0.0000            0.0000           0.000   

     nb_fungicides  nb_insecticides  nb_herbicides  mrl_fungicides  \
51               4                1              0               1   
229              0                0              0               0   
41               0                0              1               0   
151              2                0              0               0   
262              0                0              0               0   

     mrl_insecticides  mrl_herbicides  year trimester  
51                  0               0  2017         3  
229                 0               0  2019         1  
41                  0               0  2017         2  
151                 0               0  2018         3  
262                 0               0  2019         2

In [65]:
group_pesti_df.sample(5)

mrl_fungicides  mrl_insecticides  mrl_herbicides
REF....SUBSTANCE                                                  
F0041                          0                 0               0
B0142                          1                 0               0
F0005                          0                 0               0
F0020                          1                 0               0
B0140                          0                 0               0

In [66]:
cat_pesti_df.sample(5)

cat_label
REF....SUBSTANCE           
F0031                     5
F0069                     1
F0023                     1
B0037                     5
B0067                     3

##### sites

In [67]:
sites_df.sample(5)

Site  classCLC         dist       surface country
326  F0021       311   768.540427  5.352209e+05      FR
627  B0158       231   959.550638  4.010538e+05      BE
409  B0021       231  1237.725789  2.695177e+05      BE
335  F0021       242   736.016579  2.754422e+06      FR
317  F0017       242   501.012115  1.462918e+06      FR

In [68]:
by_sites_df.sample(5)

Site         111           112            121  122  123  124  131  133  \
23  B0041  391.077976  2.904582e+06     154.142379  0.0  0.0  0.0  0.0  0.0   
53  F0013    0.000000  2.320996e+03       0.000000  0.0  0.0  0.0  0.0  0.0   
43  B0105    0.000000  1.102039e+03  115351.826583  0.0  0.0  0.0  0.0  0.0   
0   F0069    0.000000  1.442780e+03  358708.730957  0.0  0.0  0.0  0.0  0.0   
59  B0069    0.000000  7.328884e+03       0.000000  0.0  0.0  0.0  0.0  0.0   

    141  ...          311          312  313         322       324       331  \
23  0.0  ...     0.000000     0.000000  0.0    0.000000   0.00000  0.970683   
53  0.0  ...  2226.353806  1645.072615  0.0  485.797839  89.28406  0.000000   
43  0.0  ...     0.000000     0.000000  0.0    0.000000   0.00000  0.000000   
0   0.0  ...   352.189664     0.000000  0.0    0.000000   0.00000  0.000000   
59  0.0  ...   201.799909     0.000000  0.0    0.000000   0.00000  0.000000   

    412  423  511  512  
23  0.0  0.0  0.0  0.0  
53  0.0  0.0  0.0  0.0  
43  0.0  0.0  0.0  0.0  
0   0.0  0.0  0.0  0.0  
59  0.0  0.0  0.0  0.0  

[5 rows x 26 columns]

##### Merge HM

In [69]:
cat_hm_df.index.name

'REF....SUBSTANCE'

In [70]:
cat_hm_df = cat_hm_df.reset_index()

In [71]:
cat_hm_df = cat_hm_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [72]:
cat_hm_df.head(3)

Site  cat_label
0  B0012          0
1  B0017          0
2  B0019          0

In [73]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  133  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   141  ...           311  312  313  322         324  331  412  423  511  512  
0  0.0  ...    352.189664  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  24001.488078  0.0  0.0  0.0  926.217913  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  11215.516811  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  

[3 rows x 26 columns]

In [74]:
merged_hm = pd.merge(by_sites_df, cat_hm_df, on=["Site"])

In [75]:
merged_hm.sample(5)

Site  111           112         121         122  123  124         131  \
7   B0061  0.0  1.225678e+02    0.000000    0.000000  0.0  0.0    0.000000   
38  B0017  0.0  7.607874e+03    2.319998    0.000000  0.0  0.0    0.000000   
33  F0029  0.0  2.793964e+02    0.000000    0.000000  0.0  0.0  751.937308   
57  B0152  0.0  4.575279e+06  522.788005  421.402885  0.0  0.0    0.000000   
1   F0025  0.0  0.000000e+00    0.000000  700.801763  0.0  0.0    0.000000   

           133  141  ...          312           313         322         324  \
7     0.000000  0.0  ...  1424.466669  4.359061e+06    0.000000   52.940801   
38    0.000000  0.0  ...  1467.201032  2.275167e+03    0.000000    0.000000   
33  338.605936  0.0  ...     0.000000  0.000000e+00    0.000000    0.000000   
57    0.000000  0.0  ...     0.000000  0.000000e+00  451.314675    0.000000   
1     0.000000  0.0  ...     0.000000  0.000000e+00    0.000000  926.217913   

    331  412  423  511  512  cat_label  
7   0.0  0.0  0.0  0.0  0.0          0  
38  0.0  0.0  0.0  0.0  0.0          0  
33  0.0  0.0  0.0  0.0  0.0          0  
57  0.0  0.0  0.0  0.0  0.0          0  
1   0.0  0.0  0.0  0.0  0.0          0  

[5 rows x 27 columns]

In [76]:
merged_hm.shape

(60, 27)

#### Merge Pesticides

In [77]:
cat_pesti_df.index.name

'REF....SUBSTANCE'

In [78]:
cat_pesti_df = cat_pesti_df.reset_index()

In [79]:
cat_pesti_df = cat_pesti_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [80]:
cat_pesti_df.head(3)

Site  cat_label
0  B0012          3
1  B0017          3
2  B0019          0

In [81]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  133  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   141  ...           311  312  313  322         324  331  412  423  511  512  
0  0.0  ...    352.189664  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  24001.488078  0.0  0.0  0.0  926.217913  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  11215.516811  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  

[3 rows x 26 columns]

In [82]:
merged_pesti = pd.merge(by_sites_df, cat_pesti_df, on=["Site"])

In [83]:
merged_pesti.sample(5)

Site  111         112  121         122  123  124            131  133  \
55  F0003  0.0  385.967531  0.0    0.000000  0.0  0.0       0.000000  0.0   
1   F0025  0.0    0.000000  0.0  700.801763  0.0  0.0       0.000000  0.0   
56  B0158  0.0  157.450834  0.0  392.706036  0.0  0.0       0.000000  0.0   
6   B0061  0.0  122.567799  0.0    0.000000  0.0  0.0       0.000000  0.0   
48  F0015  0.0    0.000000  0.0    0.000000  0.0  0.0  309985.414551  0.0   

    141  ...          312           313  322         324  331  412  423  511  \
55  0.0  ...   948.728456  5.391036e+03  0.0    0.000000  0.0  0.0  0.0  0.0   
1   0.0  ...     0.000000  0.000000e+00  0.0  926.217913  0.0  0.0  0.0  0.0   
56  0.0  ...     0.000000  0.000000e+00  0.0    0.000000  0.0  0.0  0.0  0.0   
6   0.0  ...  1424.466669  4.359061e+06  0.0   52.940801  0.0  0.0  0.0  0.0   
48  0.0  ...  2610.097612  0.000000e+00  0.0    0.000000  0.0  0.0  0.0  0.0   

    512  cat_label  
55  0.0          3  
1   0.0          0  
56  0.0          3  
6   0.0          0  
48  0.0          3  

[5 rows x 27 columns]

In [84]:
merged_pesti.shape

(57, 27)

##### Merge All

In [85]:
group_hm_df.index.name

'REF....SUBSTANCE'

In [86]:
group_pesti_df.index.name

'REF....SUBSTANCE'

In [87]:
group_hm_df = group_hm_df.reset_index()
group_pesti_df = group_pesti_df.reset_index()

In [88]:
group_hm_df = group_hm_df.rename(columns={"REF....SUBSTANCE": "Site"})
group_pesti_df = group_pesti_df.rename(columns={"REF....SUBSTANCE": "Site"})

In [89]:
by_sites_df.head(3)

Site  111          112            121         122  123  124  131  133  \
0  F0069  0.0  1442.780424  358708.730957    0.000000  0.0  0.0  0.0  0.0   
1  F0025  0.0     0.000000       0.000000  700.801763  0.0  0.0  0.0  0.0   
2  F0027  0.0   260.416070       0.000000    0.000000  0.0  0.0  0.0  0.0   

   141  ...           311  312  313  322         324  331  412  423  511  512  
0  0.0  ...    352.189664  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  
1  0.0  ...  24001.488078  0.0  0.0  0.0  926.217913  0.0  0.0  0.0  0.0  0.0  
2  0.0  ...  11215.516811  0.0  0.0  0.0    0.000000  0.0  0.0  0.0  0.0  0.0  

[3 rows x 26 columns]

In [90]:
group_hm_df.head(3)

Site  mrl_lead  mrl_cadmium  mrl_arsenic  mrl_mercury  mrl_copper
0  B0012         0            0            0            0           0
1  B0017         0            0            0            0           0
2  B0019         0            0            0            0           0

In [91]:
group_pesti_df.head(3)

Site  mrl_fungicides  mrl_insecticides  mrl_herbicides
0  B0012               1                 0               0
1  B0017               1                 0               0
2  B0019               0                 0               0

In [92]:
cat_hm_df = cat_hm_df.rename(columns={"cat_label": "hm_cat_label"})
cat_hm_df.head(3)

Site  hm_cat_label
0  B0012             0
1  B0017             0
2  B0019             0

In [93]:
cat_pesti_df = cat_pesti_df.rename(columns={"cat_label": "pesti_cat_label"})
cat_pesti_df.head(3)

Site  pesti_cat_label
0  B0012                3
1  B0017                3
2  B0019                0

In [94]:
merged_all = pd.merge(by_sites_df, group_hm_df, on=["Site"])
merged_all = pd.merge(merged_all, cat_hm_df, on=["Site"])
merged_all = pd.merge(merged_all, group_pesti_df, on=["Site"])
merged_all = pd.merge(merged_all, cat_pesti_df, on=["Site"])

In [95]:
merged_all.shape

(57, 36)

### Final Cleaning

##### HM

In [96]:
merged_hm = merged_hm.drop(['Site'], axis=1)

In [97]:
merged_pesti.sample(5)

Site          111          112           121          122  123  \
26  F0055     0.000000  1313.065065  2.656270e+03     0.000000  0.0   
31  F0033     0.000000    22.019966  0.000000e+00     0.000000  0.0   
14  F0043     0.000000  1502.358180  8.333661e+02     0.000000  0.0   
17  B0140  1416.095618  2959.662656  1.258021e+06  1967.770812  0.0   
11  F0019     0.000000  3367.677891  4.276874e+02   334.489944  0.0   

            124         131  133          141  ...          312          313  \
26     0.000000  121.787548  0.0     0.000000  ...     0.000000     0.000000   
31     0.000000    0.000000  0.0     0.000000  ...  2274.376488  2808.259665   
14  1557.870715    0.000000  0.0     0.000000  ...     0.000000     0.000000   
17     0.000000    0.000000  0.0  2481.457626  ...     0.000000     0.000000   
11     0.000000    0.000000  0.0     0.000000  ...     0.000000     0.000000   

           322          324  331  412  423         511  512  cat_label  
26    0.000000  2812.564594  0.0  0.0  0.0    0.000000  0.0          3  
31  700.999059     0.789598  0.0  0.0  0.0    0.000000  0.0          0  
14    0.000000     0.000000  0.0  0.0  0.0    0.000000  0.0          0  
17    0.000000     0.000000  0.0  0.0  0.0  544.152451  0.0          0  
11    0.000000     0.000000  0.0  0.0  0.0    0.000000  0.0          3  

[5 rows x 27 columns]

In [98]:
merged_hm.to_csv('merged_hm.csv')

##### Pesticides

In [99]:
merged_pesti = merged_pesti.drop(['Site'], axis=1)

In [100]:
merged_pesti.sample(5)

111           112           121          122  123  124  131  133  \
44     0.000000  3.399100e+02  0.000000e+00     0.000000  0.0  0.0  0.0  0.0   
11     0.000000  3.367678e+03  4.276874e+02   334.489944  0.0  0.0  0.0  0.0   
17  1416.095618  2.959663e+03  1.258021e+06  1967.770812  0.0  0.0  0.0  0.0   
10   209.227976  6.653099e+03  1.764701e+04     0.000000  0.0  0.0  0.0  0.0   
9      0.000000  4.836129e+06  3.770519e+02   546.163532  0.0  0.0  0.0  0.0   

            141          142  ...  312  313  322  324  331  412  423  \
44     0.000000     0.000000  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
11     0.000000  4940.312694  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
17  2481.457626     0.000000  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
10     0.000000     0.000000  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   
9   1685.864474   112.010107  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0   

            511  512  cat_label  
44     0.000000  0.0          3  
11     0.000000  0.0          3  
17   544.152451  0.0          0  
10  8800.306670  0.0          3  
9      0.000000  0.0          0  

[5 rows x 26 columns]

In [101]:
merged_pesti.to_csv('merged_pesti.csv')

##### All

In [102]:
merged_all = merged_all.drop(['Site'], axis=1)

In [103]:
merged_all.sample(5)

111           112          121         122  123  124  131  133  \
37      0.000000  0.000000e+00     0.000000    0.000000  0.0  0.0  0.0  0.0   
16  15107.362039  5.058683e+04  1616.543308   19.130127  0.0  0.0  0.0  0.0   
34      0.000000  0.000000e+00     0.000000    0.000000  0.0  0.0  0.0  0.0   
54      0.000000  4.575279e+06   522.788005  421.402885  0.0  0.0  0.0  0.0   
24      0.000000  7.916798e+02   533.791690  468.730666  0.0  0.0  0.0  0.0   

          141         142  ...  mrl_lead  mrl_cadmium  mrl_arsenic  \
37   0.000000    0.000000  ...         0            0            0   
16  78.805538    0.000000  ...         1            0            0   
34   0.000000  159.015078  ...         0            1            0   
54   0.000000    0.000000  ...         0            0            0   
24   0.000000    0.000000  ...         0            0            0   

    mrl_mercury  mrl_copper  hm_cat_label  mrl_fungicides  mrl_insecticides  \
37            0           0             0               0                 0   
16            0           0             4               1                 0   
34            0           0             3               0                 0   
54            0           0             0               1                 0   
24            0           0             0               1                 0   

    mrl_herbicides  pesti_cat_label  
37               0                0  
16               0                3  
34               0                0  
54               0                3  
24               0                3  

[5 rows x 35 columns]

In [104]:
merged_all.to_csv('merged_all.csv')

# Questions to ask

- When NaN => information irrelevant or not analysed?
- Difference France/Belgium?
- Is the fact that an insecticide is unauthorized relevant? (=> not the same law in France/Belgium)
- Or do we have to show that even banned pesticides are still found?
- How do we treat other categorical pesticides?
- How to increase the number of datas?

### Bo-bo-bonus

- FR  :  36 beehives for 335 zones
- BE  :  27 beehives for 297 zones

Here the CLC zone are left separate, could group some to test